src/v/pandaproxy/kafka_client_cache.h

```diff
@@ -59,6 +62,17 @@ class kafka_client_cache {
     size_t size() const;
     size_t max_size() const;
 
+    template<std::invocable<kafka_client_cache&> Func>
+    auto invoke_on_cache(Func&& func) {
+        return ss::try_with_gate(
+          _gc_gate, [this, func{std::forward<Func>(func)}]() mutable {
+              return _gc_lock.with(
+                [this, func{std::forward<Func>(func)}]() mutable {
+                    return std::forward<Func>(func)(*this);
+                });
+          });
```

Comment:
I see. Because it's sync, and the way that cleanup is formulated to be safe across suspension points, it's safe to call either at any time.

Reply:
> > I'm not sure I understand the function of `_gc_lock` and `_gc_gate`, the name implies they're for garbage collection, which makes me think they should be protecting the `_gc_timer`, and `clean_stale_clients`?
> > I'm not sure why the lock is used here.
> 
> Yes I can change the name but the reason is to have the gate and lock here is to protect inflight requests and `fetch_or_insert` is **note** async. So the next best place is to put the concurrency mechs here.

If it's not async, it doesn't need a gate or lock, right?

> > There's only really one caller of this function now, and just calls into `fetch_or_insert`. I guess both `fetch_or_insert` and `clean_stale_clients` mess with `_cache`, so maybe the goal is to protect that? Should the lock be moved into `clean_stale_clients` and `fetch_or_insert`? In which case `_gc_lock` could be renamed to `_cache_lock`?
> > Oh I see, `fetch_or_insert` isn't async, so you can't protect it with the lock.
> > In that case should it be made private along with lots of other methods that interact with the cache?
> 
> Yes a better name is _cache_lock and _cache_gate. I'll make that change. I don't think making fetch_of_insert private is necessary here since other places need to call that method (e.g., auth_ctx_server). I also expect to use this class in the Schema Registry whenever we introduce ACLs.

I think I finally understand the class now, `fetch_or_insert` doesn't need any protection with a lock or gate. I think invoke_on_cache isn't needed, it's just confusing.

> Sure we could make friend classes but that seems like a vulnerability. Let's chat about this more because perhaps I have an obsolete view of friend classes.

I was thinking that `invoke_on_cache` was called to protect `fetch_or_insert`, but it's not needed, it confused me.